### ISSUE TITLE
GetSystemUpStreamProxy not working

### ISSUE THREAD

**Issue by honfika** [justcoding121/Titanium-Web-Proxy#264]
Before starting TWP my proxy setting is using a PAC file: http://localhost/proxy.pac

Content is:
```
function FindProxyForURL(url, host) {
	if (shExpMatch(host, "index.hu"))
	{
		return "PROXY 127.0.0.1:8888";
	}

	if (shExpMatch(host, "*.index.hu"))
	{
		return "PROXY 127.0.0.1:8888";
	}

	return "DIRECT";
}
```

On port 8888 a concurrent proxy is running, not registered as a system proxy (only inside the PAC file). It has no upstream proxy (not using the system proxy, it would be a loop).

`GetSystemUpStreamProxy` tries to resolve the upstream proxy using the `WebClient` class, so for testing I tried the following:
```
            var systemProxyResolver1 = new WebProxy();
            var systemProxyResolver2 = WebProxy.GetDefaultProxy();
            Console.WriteLine("proxy1a: " + systemProxyResolver1.GetProxy(new Uri("http://index.hu")));
            Console.WriteLine("proxy2a: " + systemProxyResolver2.GetProxy(new Uri("http://index.hu")));

            proxyServer.SetAsSystemProxy(explicitEndPoint, ProxyProtocolType.AllHttp);

            var systemProxyResolver3 = new WebProxy();
            var systemProxyResolver4 = WebProxy.GetDefaultProxy();

            Console.WriteLine("proxy1b: " + systemProxyResolver1.GetProxy(new Uri("http://index.hu")));
            Console.WriteLine("proxy2b: " + systemProxyResolver2.GetProxy(new Uri("http://index.hu")));
            Console.WriteLine("proxy3: " + systemProxyResolver3.GetProxy(new Uri("http://index.hu")));
            Console.WriteLine("proxy4: " + systemProxyResolver4.GetProxy(new Uri("http://index.hu")));

```
Result:
```
proxy1a: http://index.hu/
proxy2a: http://127.0.0.1:8888/
Set endpoint at Ip 0.0.0.0 and port: 8000 as System HTTP and HTTPS Proxy
proxy1b: http://index.hu/
proxy2b: http://127.0.0.1:8000/
proxy3: http://index.hu/
proxy4: http://127.0.0.1:8000/

```

Conclusion:
- `new WebProxy()` always returns the original URL, which is absolutely wrong. (proxy1a, 1b, 3)
- The current code in `GetSystemUpStreamProxy` does not work in any case. Do you have other expreience? Was it working earlier?
- Does not matter when was the `WebProxy` object created. (1b=3 and 2b=4)

The only correct result is 2a, which was returned when TWP was not registered as system proxy.
So when TWP is registered as a system proxy it does not work in any way.
WebProxy.GetDefaultProxy is deprecated also...

Probably the solution would be be using the `WinHttpDetectAutoProxyConfigUrl` native WinHTTP function: https://msdn.microsoft.com/en-us/library/windows/desktop/aa384094(v=vs.85).aspx

**Comment by justcoding121** [comment]
I believe @aricih contributed code related to this. I have'nt done any testing on it.

If  `WebProxy.GetDefaultProxy();` works at least when System proxy is not set we can use it for now. You said after setting system proxy the call did not work. But in you example below the result is same as proxy2a. Is that expected?

`Console.WriteLine("proxy4: " + systemProxyResolver4.GetProxy(new Uri("http://index.hu")));`

**Comment by justcoding121** [comment]
Oh. So the port 8000 is wrong.

**Comment by honfika** [comment]
Yes, port 8000 is wrong, it is the port of TWP proxy. I'd like to fix this issue, but it needs more time. I'll be on holiday from Thursday, I can check it in the following 2 weeks.

**Comment by justcoding121** [comment]
Thanks for the new PR. 

One question did you try #270 for the above test cases? I noted he used `WebRequest.GetSystemWebProxy()` instead of `WebProxy.GetSystemUpStreamProxy()`. If that still does'nt fix the problem, then we can merge this new PR. I like to get your thoughts.

**Comment by honfika** [comment]
I tried it now. The result is exactly the same as with `WebProxy.GetDefaultProxy()`.
So it works only when TWP is not registered as a system proxy.
When it is registered, it returns the address of TWP. (Even when I cache the `IWebProxy` object)
But at least it is not deprecated as `WebProxy.GetDefaultProxy()`

Could you please wait a little bit? I'm adding some tests.

**Comment by justcoding121** [comment]
Sure

**Comment by honfika** [comment]
I've added the tests.

**Comment by justcoding121** [comment]
I tried latest develop after the merge. Noted that connections keep increasing exponentially, hitting as much as 2000 server connections. Something is wrong. I am disabling the `ForwardToUpstreamGateway` in example project until we figure it out.

**Comment by justcoding121** [comment]
![image](https://user-images.githubusercontent.com/2637679/27243489-202c0e1e-52b0-11e7-86e3-9fcd52d8782c.png)

Looks like an infinite loop. I noted that below code is called and returns the address of our proxy.

` var proxy = systemProxyResolver.GetProxy(sessionEventArgs.WebSession.Request.RequestUri);`

If you going to make a patch can you also update the proxy setting restore functionality so that if the existing address was same as our proxy server listening address, then clear the proxy settings upon quit. This will be helpful to avoid manually clearing the system proxy if proxy crashes abruptly and later if I restart and quit the proxy.

**Comment by honfika** [comment]
Hmm, what are the proxy settings when you start TWP?
It it is no proxy there is no problem.

**Comment by justcoding121** [comment]
Okay so that's the problem. I terminated debugging from visual studio, which was not a gracious shut down. Subsequently if I start proxy the` GetProxy` call returns the same address as our proxy listening address. To avoid this we need to check if the returned proxy address is same as our listening address and if so return null.

Also like I mentioned above we don't need to 'restore' system proxy settings if the restore address is same as our listening address. Since after quit nobody is listening in that restore address, we should be safe to reset system proxy settings to nothing.
